pennylane_lightning/lightning_qubit.py

```diff
@@ -226,6 +394,33 @@ def apply_lightning(self, state, operations):
 
         return np.reshape(state_vector, state.shape)
 
+    def apply(self, operations, rotations=None, **kwargs):
+        # State preparation is currently done in Python
+        if operations:  # make sure operations[0] exists
+            if isinstance(operations[0], QubitStateVector):
+                self._apply_state_vector(operations[0].parameters[0].copy(), operations[0].wires)
+                del operations[0]
+            elif isinstance(operations[0], BasisState):
+                self._apply_basis_state(operations[0].parameters[0], operations[0].wires)
+                del operations[0]
+
+        for operation in operations:
+            if isinstance(operation, (QubitStateVector, BasisState)):
+                raise DeviceError(
+                    "Operation {} cannot be used after other Operations have already been "
+                    "applied on a {} device.".format(operation.name, self.short_name)
+                )
+
+        if operations:
+            self._pre_rotated_state = self.apply_lightning(self._state, operations)
+        else:
+            self._pre_rotated_state = self._state
+
+        if rotations:
+            self._state = self.apply_lightning(np.copy(self._pre_rotated_state), rotations)
```

Comment:
I tried to get rid of the copy during development, but in the way that `apply_lightning` is coded we need to provide a copy of `_pre_rotated_state`.
In `apply_lightning`, the inputted state and the outputted state_vector share the same pointer.
[Here](https://github.com/PennyLaneAI/pennylane-lightning/blob/device-inheritance/pennylane_lightning/lightning_qubit.py#L362) is where it happens. 
I didn't think too much about that, because this is the way that was before.

Reply:
For now, I see no problem with how `apply_lightning` is coded, though.

Reply:
Ok, let's call this good for now. I suspect we can always try to improve things at a later date if necessary.